File: prot1_full_ai_deepseek/decomp_report/ghidra_script/ExportDecomp.java
// ExportDecomp.java — Ghidra headless post-script (Java).
// Dumps every function's name/entry/size, decompiler CFG block+edge counts,
// and full decompiled C to the output file (first script arg).
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.pcode.PcodeBlockBasic;

import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;

public class ExportDecomp extends GhidraScript {

    @Override
    public void run() throws Exception {
        String outpath = (getScriptArgs().length > 0) ? getScriptArgs()[0] : "/tmp/decomp.txt";
        FunctionManager fm = currentProgram.getFunctionManager();
        DecompInterface decomp = new DecompInterface();
        decomp.openProgram(currentProgram);

        List<Function> funcs = new ArrayList<>();
        for (Function f : fm.getFunctions(true)) {
            funcs.add(f);
        }
        funcs.sort((a, b) -> Long.compare(a.getEntryPoint().getOffset(), b.getEntryPoint().getOffset()));

        StringBuilder sb = new StringBuilder();
        for (Function fn : funcs) {
            sb.append("===== FUNCTION ").append(fn.getName())
              .append(" @ 0x").append(Long.toHexString(fn.getEntryPoint().getOffset()))
              .append(" size=").append(fn.getBody().getNumAddresses()).append(" =====\n");
            DecompileResults res = decomp.decompileFunction(fn, 60, monitor);
            if (res != null && res.decompileCompleted() && res.getHighFunction() != null) {
                List<PcodeBlockBasic> bbs = res.getHighFunction().getBasicBlocks();
                int blocks = bbs.size();
                int edges = 0;
                for (PcodeBlockBasic bb : bbs) {
                    edges += bb.getOutSize();
                }
                sb.append("CFG blocks=").append(blocks).append(" edges=").append(edges).append("\n");
                sb.append(res.getDecompiledFunction().getC()).append("\n");
            } else {
                sb.append("(decompile failed)\n");
            }
        }

        try (PrintWriter pw = new PrintWriter(outpath)) {
            pw.print(sb.toString());
        }
        println("wrote " + outpath);
    }
}

File: evaluation/decomp_report/ghidra_script/export_decomp.py
# export_decomp.py — Ghidra headless post-script.
# Dumps, for every function in the program:
#   - name, entry address, size
#   - basic-block / edge counts from the decompiler CFG model
#   - the full decompiled C
# Usage: analyzeHeadless <proj> <projname> -import <bin> -scriptPath <dir> \
#            -postScript export_decomp.py <outfile>
import sys

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

outpath = args[0] if len(args) > 0 else "/tmp/decomp.txt"
monitor = ConsoleTaskMonitor()
program = currentProgram
fm = program.getFunctionManager()

decomp = DecompInterface()
decomp.openProgram(program)

lines = []
fm = program.getFunctionManager()
funcs = sorted(fm.getFunctions(True), key=lambda fn: fn.getEntryPoint().getOffset())
for fn in funcs:
    name = fn.getName()
    entry = fn.getEntryPoint().getOffset()
    body = fn.getBody()
    size = body.getNumAddresses() if body else 0
    lines.append("===== FUNCTION %s @ 0x%x size=%d =====" % (name, entry, size))
    res = decomp.decompileFunction(fn, 60, monitor)
    if res is not None and res.decompileCompleted():
        ccm = res.getCCodeModel()
        if ccm is not None:
            bbs = ccm.getBasicBlocks()
            nedges = 0
            for bb in bbs:
                nedges += len(bb.getOutEdges())
            lines.append("CFG blocks=%d edges=%d" % (len(bbs), nedges))
        cfn = res.getDecompiledFunction()
        if cfn is not None:
            lines.append(cfn.getC())
    else:
        lines.append("(decompile failed)")

with open(outpath, "w") as f:
    f.write("\n".join(lines))
    f.write("\n")

print("wrote " + outpath)